PySCF Driver Updates

mrossinek opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

This PR continues the work #166 by adding general improvements to the `PySCFDriver`. The others drivers will need to be tackled one-by-one in a similar fashion.

### Details and comments

- [x] integration of `integrals.py` as class methods
- [x] improved splitting of class methods into more logical subunits
- [x] updated docstrings
- [x] closes #28:
  - explicit checking of all input arguments is _not_ feasible
  - unfortunately, PySCF also does not provide simple means to validate the arguments through some method call
  - BUT, through more logical method splitting, a more meaningful error will now be raised (e.g. [this one](https://github.com/mrossinek/qiskit-nature/blob/41b2945ea572e355aace5e877a135c4a538cf904/qiskit_nature/drivers/pyscfd/pyscfdriver.py#L371))
- [x] also re-evaluates #29 for this driver by updating all links to the new PySCF website (https://pyscf.org)
- [x] update unittests 


:warning: This PR as significant conflicts with #173!

mrossinek commented:
@woodsp-ibm While this is still a WIP, it would be great if you could already take a look at this and provide me some early feedback :+1:

woodsp-ibm commented on test/drivers/pyscfd/test_driver_problem_pyscf.py:
```diff
@@ -122,6 +123,7 @@ def test_lih_rhf_bk(self):
         )
         self._assert_energy_and_dipole(result, "lih")
 
+    @slow_test
```
I am not sure these are all that slow. I would be reluctant not to have these in there running normally under CI.

mrossinek replied:
I do not see the benefit of the `OH` tests compared to the existing `H2` *and* `LiH` tests. The `OH` tests are, relative to the other ones, *very* slow. Hence, I disabled them. If you disagree strongly we can re-enable them. However, for the record, I don't think they are necessary.

woodsp-ibm replied:
The doublet OH was included since its not a closed shell molecule to start with unlike LiH. Its a bit bigger and hence the bit slower. It was to make sure we were doing things right with either type of molecule across the different methods - it was originally called test driver methods. How long does it take to run that you consider it slow? If anything I could support perhaps putting some of them oh_uhf/oh_rohf (same for lih_uhf/rohf/rhf) as slow where they test the same thing but with a different mapping to thin things down. Arguably if it works for one mapping it should be good for the others - if it does not its more likely the mapper is broke than the driver. Having said that I am not sure we have mapping tests with such coverage of open shell molecules elsewhere.

woodsp-ibm commented on qiskit_nature/drivers/pyscfd/pyscfdriver.py:
```diff
@@ -53,47 +68,79 @@ def __init__(
+                etc. Note, that more advanced configuration options like a Dictionary or custom
+                basis sets are not allowed for the moment. Refer to 4_ for an extensive list of
+                PySCF's valid basis set names.
+            hf_method: The SCF method type to be used for the PySCF calculation. While the name
+                refers to HF methods, the PySCFDriver also supports KS methods. Refer to the
+                ``HFMethodType`` for a list of the supported methods.
```
If the name needs this further explanation around hf would it be better to have this as a different name? Maybe simply just `method` that accommodates either and deprecate the `hf_method`; same for HFMethodType maybe it needs an enum specific to PySCF - that one supported  all the drivers and was used in PyQuante where these additional methods would be invalid.

mrossinek replied:
I agree that the naming is not ideal. Given our plans for the `MoleculeDriver` I think we can tackle that then, because we need to re-evaluate how this Enum ties into the general case there, too.

woodsp-ibm commented on qiskit_nature/drivers/pyscfd/pyscfdriver.py:
```diff
@@ -125,49 +302,386 @@ def _check_valid():
+
+        self._calc = hf_method_cls(self._mol)
+
+        if "KS" in hf_method_name:
```
Since we have enum here we should just be able to ask if the enum is one of the set of them that should be acted upon - that way we don't have some string "KS" as comparison that is we changed enum names, with set the refactoring would tale care of things.

mrossinek replied:
The reason why I went with this comparison style is that I did not to duplicate half of the Enum as a set here. And this comparison properly reflects the nature of the check which we care about.

woodsp-ibm replied:
I guess in my mind it makes the code 'brittle' in that refactoring an enum, should we ever want to change any name, this detection here would break. Listing the set and simply saying `if x in [A, B, C]` works and if we refactor C to D that code should get changed by the refactor whereas some check on "C" would not.

woodsp-ibm commented on qiskit_nature/drivers/pyscfd/pyscfdriver.py:
```diff
@@ -125,49 +302,386 @@ def _check_valid():
-            "",
         ]
-        q_mol.origin_driver_config = "\n".join(cfg)
+        if "ks" in self._hf_method.lower():
```
I guess this "ks" string detection is lower case versus the above I commented on where it was uppercase; again with an enum I think we should be able to leverage that as described before

woodsp-ibm commented on qiskit_nature/drivers/pyscfd/pyscfdriver.py:
```diff
@@ -125,49 +302,386 @@ def _check_valid():
+            QiskitNatureError: If an invalid HF method type was supplied.
+        """
+        try:
+            hf_method_name = self._hf_method.upper()
+            hf_method_cls = getattr(scf, hf_method_name)
```
I guess this works - but it ends up more obscure for me and could do with explanation for any future readers if you choose to want to keep this. Certainly the switch/case (ifs from before) would grow larger here but to me its clearer/cleaner - now the enum names I guess have to match the method names.

mrossinek replied:
I can add some explanatory comment. I will not revert it to the `if-elif-else` switches.
I will wait with this update until we have resolved the other discussion points above.

woodsp-ibm replied:
Thats fine - if its documented - though personally I am not so keen on the direct couping of names in the enum - that currently has wider use in Nature, to names that match PySCF driver methods, though I understand RHF/ROHF and UHF are universal enough - maybe that's true of these KS methods too.

mrossinek commented:
I will try to tackle @woodsp-ibm comments later today or early tomorrow.
However, I strongly believe we should merge #173 first. I will ping that one again now

mrossinek commented:
I just rebased this branch after the various updates to the `PySCFDriver` and its relocation. I also addressed @woodsp-ibm's comments so hopefully this is now entirely ready for review :+1:

woodsp-ibm commented on qiskit_nature/drivers/second_quantization/pyscfd/pyscfdriver.py:
```diff
@@ -10,25 +10,39 @@
+
+    warnings.filterwarnings("ignore", category=DeprecationWarning, module="pyscf")
+except ImportError:
+    logger.info("PySCF is not installed. See https://sunqm.github.io/pyscf/install.html")
```
The former code was fixed with an updated message for the changed install location as follows
```suggestion
    logger.info("PySCF is not installed. See https://pyscf.org/install.html")
```

mrossinek replied:
Thanks! I must have missed this one. After fixing it grepping for `sunqm` no longer returns any results in the codebase :+1:

woodsp-ibm commented on test/drivers/second_quantization/gaussiand/test_driver_methods_gaussian.py:
```diff
@@ -10,18 +10,18 @@
 
-class TestDriverMethodsGaussian(TestDriverMethods):
-    """Driver Methods Gaussian tests"""
+class TestDriverProblemGaussian(TestDriverProblem):
```
The original rationale why this was called TestDriverMethods was that it was really around tests for the different methods - rhf, rohf, uhf

mrossinek replied:
I see. I renamed it because these tests are now all based on the `Problem` classes whereas the other drivers tests are more basic. I don't think it matters too much though.

woodsp-ibm replied:
They were named based on _what_ they were testing not _how_ - i.e. it's not about the problem, that's just a means to an end to verify the function/result. If we could exact diagonalize a FermionicOp to get a min eig value it could be done that way given what is tested perhaps - before we used to use core Hamiltonian and Numpy - using Problem and GroundStateSolver is just the latest way.

mrossinek replied:
After an offline-sync and re-evaluating my view on these tests I have decided to revert this naming change. Initially I was thinking differently about this but I agree that this is a change which does not add any value. I will revert this later today.

mrossinek commented:
@pbark @woodsp-ibm A fresh review on this PR would also be great :upside_down_face:

mrossinek commented:
All comments (here and offline) should be resolved now, @woodsp-ibm @pbark

pbark reviewed (approved):
Thanks @mrossinek! LGTM!
